### ISSUE TITLE
Certificate Name Mismatch When Establishing Connection Through Proxy

### ISSUE THREAD

**Issue by joellubi-cherre** [GoogleCloudPlatform/alloydb-auth-proxy#111]
## Bug Description

I am testing the AlloyDB Auth Proxy, using `psql` to query the database. I'm running the proxy in a GKE instance in the same VPC as the AlloyDB instance. Specifically, I am running the `gcr.io/alloydb-connectors/alloydb-auth-proxy` image and the `postgres` image on the same Pod and making the connection via localhost.

2 days ago I began my testing and was able to successfully perform all actions as expected. Yesterday I did not connect to the instance, but today I began my testing again. However, I have been unable to connect and there appears to be a mismatch between certificate names.

When connecting with `psql`, I get the following error:
```
root@alloydb-hasura-8977b4c5d-9mhf4:/# psql -U postgres -h localhost -p 5432
psql: error: connection to server at "localhost" (127.0.0.1), port 5432 failed: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

After attempting to make this connection, the following can be found in the alloydb auth proxy logs:
```
alloydb-auth-proxy 2022/08/26 19:20:30 Authorizing with Application Default Credentials
alloydb-auth-proxy 2022/08/26 19:20:30 [**REDACTED__INSTANCE_URI**] Listening on 127.0.0.1:5432
alloydb-auth-proxy 2022/08/26 19:20:30 The proxy has started successfully and is ready for new connections!
alloydb-auth-proxy 2022/08/26 19:21:18 [**REDACTED__INSTANCE_URI**] accepted connection from 127.0.0.1:52770
alloydb-auth-proxy 2022/08/26 19:21:18 [**REDACTED__INSTANCE_URI**] failed to connect to instance: Dial error: handshake failed (instance URI = "**REDACTED__INSTANCE_URI_SHORT_FORM**"): Dial error: certificate had CN "**REDACTED__NOT_THE_UID_OF_INSTANCE**.server.alloydb", expected "**REDACTED__UID_OF_INSTANCE**.server.alloydb" (instance URI = "**REDACTED__INSTANCE_URI_SHORT_FORM**")
```

## Example code (or command)

Here is the relevant subset of the Kubernetes PodSpec
```
containers:
    - name: alloydb-auth-proxy
       image: gcr.io/alloydb-connectors/alloydb-auth-proxy
       command:
         - "/alloydb-auth-proxy"
         - "**REDACTED__INSTANCE_URI**"

    - name: postgres
       image: postgres
       command: ['/bin/sh', '-c', 'while true; do sleep 100; done;']
```

## How to reproduce

_I don't know that the following will reliably reproduce the issue but this has been my experience_

  1. Start alloydb auth proxy and successfully connect to instance using it.
  2. Wait more than 24 hrs (I only mention this because it may be related to certificates expiring after 24 hrs as mentioned [here](https://cloud.google.com/alloydb/docs/auth-proxy/overview#how-works); may or may not be related).
  3. Start alloydb auth proxy and attempt to connect again.

## Environment

1. OS type and version: GKE Node
2. Proxy version: `gcr.io/alloydb-connectors/alloydb-auth-proxy` (Image tag was not specified at any point, so I'm not certain the versions pulled were the same)

**Comment by deepankarjha** [comment]
Facing the same issue here.

**Comment by Vincecoquart** [comment]
I ve got the same issue. Last week, I recreated the cluster several times (4 times). The proxy worked with each new installation. But after a certain period, the issue happens again. And i vent found a way to fix it.. The issue is not ephemeral, it happens once i ve observed it once

**Comment by joellubi-cherre** [comment]
Hi @kurtisvg, thanks for your response.

Our experience has been similar to that of others in this thread.

1. The cluster is almost 2 weeks old. This is the only one I've tested the auth proxy on.
2. Yes, we could connect to the cluster using the auth proxy without issues before.
3. Once it has been observed once, it persists indefinitely.

One other potentially relevant detail is that we are still able to connect directly via private IP of the instance. This is adequate for some of the tests we are currently running but we will need the auth proxy working before we can migrate any production workloads.

**Comment by kurtisvg** [comment]
Thanks for the info folks. We will follow up internally and update here if we need anything

**Comment by enocom** [comment]
The backend team has identified the problem and is actively working on a fix. I'll update here as I have more to share.

**Comment by enocom** [comment]
For context: the Go connector added a validation check to ensure the server side proxy's certificate matches what the API reports. That change was introduced in https://github.com/GoogleCloudPlatform/alloydb-auth-proxy/blob/e9251a2cf1102053ce70cd92ad3779bc6fecbcf0/go.mod#L6 which was released as part v0.4.0. Any prior version shouldn't have this issue.

**Comment by enocom** [comment]
The fix is complete and being deployed. I'll update here when it's out.

**Comment by enocom** [comment]
Bumping down the priority since there is at least a workaround.

**Comment by enocom** [comment]
The rollout is still in progress. Again I’ll update here when I have more information.

**Comment by cbusseata** [comment]
Hey @joellubi-cherre, I came across this closed issue while trying to troubleshoot using alloydb-auth-proxy in  GKE myself. I unfortunately have never been able to connect. I have an alloydb cluster/instance in the same VPC as my GKE cluster. However, I am getting
````
failed to connect to instance: Dial error: failed to dail (instance URI = "<INSTANCE>"): dial tcp <private IP>:5433: i/o timeout
````

I was wondering if you would share some details about your GKE configuration? By chance is yours a VPC-Native cluster? (mine is not)

**Comment by enocom** [comment]
@cbusseata Let's open a new issue and mark it as a question. We can discuss that there.
